plugin/evm/message/handshake/upgrade_config_message.go

```diff
@@ -0,0 +1,117 @@
+	// Config for enabling and disabling precompiles as network upgrades.
+	PrecompileUpgrades []rawPrecompileUpgrade `serialize:"true"`
+}
+
+type UpgradeConfigMessage struct {
+	bytes []byte
+	hash  []byte
+}
+
+func (u *UpgradeConfigMessage) Bytes() []byte {
+	return u.bytes
+}
+
+func (u *UpgradeConfigMessage) ID() []byte {
+	return u.hash
+}
+
+// Attempts to parse a networkUpgradeConfigMessage from a []byte
+//
+// This function attempts to parse a stream of bytes as a
+// networkUpgradeConfigMessage (as serialized from
+// UpgradeConfigToNetworkMessage).
+//
+// The function returns a reference of *params.UpgradeConfig
+func NewUpgradeConfigMessageFromBytes(bytes []byte) (*params.UpgradeConfig, error) {
+	var config networkUpgradeConfigMessage
+	version, err := Codec.Unmarshal(bytes, &config)
+	if err != nil {
+		return nil, err
+	}
+	if version != Version {
+		return nil, ErrInvalidVersion
+	}
+
+	var PrecompileUpgrades []params.PrecompileUpgrade
+	for _, precompileUpgrade := range config.PrecompileUpgrades {
+		module, ok := modules.GetPrecompileModule(precompileUpgrade.Key)
+		if !ok {
+			return nil, ErrUnknowPrecompile
+		}
+		preCompile := module.MakeConfig()
+		version, err := Codec.Unmarshal(precompileUpgrade.Bytes, preCompile)
+		if version != Version {
+			return nil, ErrInvalidVersion
+		}
+		if err != nil {
+			return nil, err
+		}
+		PrecompileUpgrades = append(PrecompileUpgrades, params.PrecompileUpgrade{Config: preCompile})
+	}
+
+	return &params.UpgradeConfig{
+		OptionalNetworkUpgrades: config.OptionalNetworkUpgrades,
+		StateUpgrades:           config.StateUpgrades,
+		PrecompileUpgrades:      PrecompileUpgrades,
+	}, nil
+}
+
+// Wraps an instance of *params.UpgradeConfig
+//
+// This function returns the serialized UpgradeConfig, ready to be send over to
+// other peers. The struct also includes a hash of the content, ready to be used
+// as part of the handshake protocol.
+//
+// Since params.UpgradeConfig should never change without a node reloading, it
+// is safe to call this function once and store its output globally to re-use
+// multiple times
+func NewUpgradeConfigMessage(config *params.UpgradeConfig) (*UpgradeConfigMessage, error) {
+	PrecompileUpgrades := make([]rawPrecompileUpgrade, 0)
+	for _, precompileConfig := range config.PrecompileUpgrades {
+		bytes, err := Codec.Marshal(Version, precompileConfig.Config)
+		if err != nil {
+			return nil, err
+		}
+		PrecompileUpgrades = append(PrecompileUpgrades, rawPrecompileUpgrade{
+			Key:   precompileConfig.Key(),
+			Bytes: bytes,
+		})
+	}
+
+	wrappedConfig := networkUpgradeConfigMessage{
+		OptionalNetworkUpgrades: config.OptionalNetworkUpgrades,
+		StateUpgrades:           config.StateUpgrades,
+		PrecompileUpgrades:      PrecompileUpgrades,
+	}
+
+	bytes, err := Codec.Marshal(Version, wrappedConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	hash := crypto.Keccak256(bytes)
+	return &UpgradeConfigMessage{
+		bytes: bytes,
+		hash:  hash[:8],
```

Comment:
can we send full hash? I don't think it worth saving 24 bytes. We can also convert `hash` to `common.Hash` and use `crypto.Keccak256Hash`.